fiftyone/migrations/revisions/v0_18_1.py

```diff
@@ -0,0 +1,147 @@
+    ):
+        if "name" in d:
+            existing_names.add(d["name"])
+
+        if "url_name" in d:
+            existing_url_names.add(d["url_name"])
+
+    if name is None:
+        name = _get_default_dataset_name(existing_names)
+
+    if url_name is None:
+        try:
+            # Give the user a one-time pass if they have a really long dataset
+            # name to have a URL-friendly name that's shorter
+            url_name = _to_url_name(name[: _URL_NAME_LENGTH_RANGE[1]])
+        except Exception as e:
+            old_name = name
+            name = _get_default_dataset_name(existing_names)
+            url_name = _to_url_name(name)
+
+            logger.warning(e)
+            logger.warning("Renaming dataset '%s' to '%s'", old_name, name)
+
+    if name in existing_names or url_name in existing_url_names:
+        old_name = name
+
+        # Make a valid, unique name that resembles the original
+        name = _get_default_dataset_name(existing_names)
+        name = old_name + "-RENAMED-" + name
+        name = name[-_URL_NAME_LENGTH_RANGE[1] :]
+
+        url_name = _to_url_name(name)
+
+        logger.warning(
+            "Dataset name '%s' conflicts with another dataset's name", old_name
+        )
+        logger.warning("Renaming dataset '%s' to '%s'", old_name, name)
+
+    dataset_dict["name"] = name
+    dataset_dict["url_name"] = url_name
+
+
+_SAFE_CHARS = set(string.ascii_letters) | set(string.digits) | set("-_.")
+_WHITESAPCE_CHARS = set(string.whitespace) | set("+")
+_URL_NAME_LENGTH_RANGE = (1, 100)
+
+
+def _sanitize_char(c):
+    if c in _SAFE_CHARS:
+        return c
+
+    if c in _WHITESAPCE_CHARS:
+        return "-"
+
+    return ""
+
+
+def _to_url_name(name):
+    if not isinstance(name, str):
+        raise ValueError("Expected string; found %s: %s" % (type(name), name))
+
+    if len(name) > _URL_NAME_LENGTH_RANGE[1]:
+        raise ValueError(
+            "'%s' is too long; length %d > %d"
+            % (name, len(name), _URL_NAME_LENGTH_RANGE[1])
+        )
+
+    safe = []
+    last = ""
+    for c in name:
+        s = _sanitize_char(c)
+        if s and (s != "-" or last != "-"):
+            safe.append(s)
+            last = s
+
+    url_name = "".join(safe).strip("-")
+
+    if len(url_name) < _URL_NAME_LENGTH_RANGE[0]:
+        raise ValueError(
+            "'%s' has invalid URL-friendly name '%s'; length %d < %d"
+            % (name, url_name, len(url_name), _URL_NAME_LENGTH_RANGE[0])
+        )
+
+    if len(url_name) > _URL_NAME_LENGTH_RANGE[1]:
+        raise ValueError(
+            "'%s' has invalid URL-friendly name '%s'; length %d > %d"
+            % (name, url_name, len(url_name), _URL_NAME_LENGTH_RANGE[1])
+        )
+
+    return url_name
+
+
+def _get_default_dataset_name(existing_names):
+    now = datetime.now()
+    name = now.strftime("%Y.%m.%d.%H.%M.%S")
+    if name in existing_names:
+        name = now.strftime("%Y.%m.%d.%H.%M.%S.%f")
```

Comment:
unique indeed

Reply:
FYI- I hate copying code, but we do have to do a one time copy of code whenever certain logic is required in migrations, since they have to run independent of library version.

In this case, I've copy-pasted the logic from `fiftyone.core.dataset.get_default_dataset_name()` and `fiftyone.core.utils.to_url_name()`.